airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/DestinationTask.kt

```diff
@@ -0,0 +1,121 @@
+package io.airbyte.cdk.write
+
+import kotlin.math.min
+
+/**
+ * A task that can be executed by a DestinationRunner.
+ *
+ * concurrency: reports the desired concurrency:
+ * - id: a unique identifier for the task
+ * - perSync: the maximum number of tasks that can run concurrently in a sync
+ * - perStream: the maximum number of tasks that can run concurrently for a single stream
```

Comment:
Oh interesting. So we either need to ensure all of a Stage1 is complete before Stage2 executes, or add exclusions to the concurrency guard?

At the task level this could be `Concurrency(id: String, perStream, perSync, exclusions: Set<String>)`. Ie, don't run if X is running.

Also in this model, the batches don't have guards, the tasks do. So to implement this, you either
* depart from the default interface and use a state machine
* add concurrency to the batches that get propagated up to the task level behind the scenes

Reply:
yeah. basically T+D needs an exclusive lock on the raw table, but the `copy into table` batches just need a shared lock

also derp, I'm going to mix up batches and tasks quite a lot :/

I don't know what the right option is. Warehouse destinations don't care (b/c we want to do `copy into` and `T+D` once per sync anyway). So this is mostly for postgres(/mysql/etc), which is a much more limited usecase.

adding concurrency to batches seems weird when we're already doing that on tasks (at least that's my impression, curious if you think this is a reasonable option). But we could probably have the CDK provide a default "database state machine", so we're not redoing it for every destination?

Reply:
Possible approaches:
* state machine for advanced users + interface for standard
* state machine for implementation + multiple interfaces w/ protocols for different use cases
* state machine + expressive flow semantics that are just as good (?) as an interface but more flexible?